core/chain.py

```
@@ -45,33 +164,54 @@ def add_block(self, block):
+                    }
+                    for b in self.chain
+                ],
+                "state": self.state.to_dict() if hasattr(self.state, 'to_dict') else {}
+            }
+
+        try:
+            temp_file = self._chain_file + ".tmp"
+            with open(temp_file, 'w') as f:
+                json.dump(data, f, indent=2)
+            os.replace(temp_file, self._chain_file)
+            logger.info(f"Saved chain with {len(self.chain)} blocks to {self._chain_file}")
```

Ruff (0.15.1)

[warning] Logging statement uses f-string

(G004)